spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindPotentialSecurityCheckBasedOnUntrustedSource.java

```diff
@@ -0,0 +1,242 @@
+        this.bugAccumulator = new BugAccumulator(bugReporter);
+    }
+
+    @Override
+    public void visit(JavaClass obj) {
+        isDoPrivileged = Subtypes2.instanceOf(getDottedClassName(), "java.security.PrivilegedAction")
+                || Subtypes2.instanceOf(getDottedClassName(), "java.security.PrivilegedExceptionAction");
+    }
+
+    @Override
+    public void visit(Method obj) {
+        nonFinalMethodsCalledOnParam.clear();
+        isDoPrivilegedRun = isDoPrivileged && "run".equals(getMethodName());
+    }
+
+    @Override
+    public void visit(Code obj) {
+        if (!isDoPrivilegedRun && (!getThisClass().isPublic() || !getMethod().isPublic())) {
+            return;
+        }
+        super.visit(obj);
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        bugAccumulator.reportAccumulatedBugs();
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        if (seen == Const.INVOKEVIRTUAL && getXClassOperand() != null && getXMethodOperand() != null) {
+            if (getXMethodOperand().getSignature().endsWith("V")) {
+                return;
+            }
+            OpcodeStack.Item object = stack.getStackItem(0);
+
+            if (isDoPrivilegedRun) {
+                if (!getXClassOperand().isFinal() && !getXMethodOperand().isFinal() && object.isInitialParameter()
+                        && isNestingMethodLocalVariable(object)) {
+                    addToMethodsCalledInsidePrivilegedAction(getXMethodOperand(), object);
+                }
+                return;
+            }
+
+            if (getXClass().isPublic() && getXMethod().isPublic() && !getXClassOperand().isFinal()
+                    && !getXMethodOperand().isFinal() && object.isInitialParameter() && object.getXField() == null
+                    && topParameterName != null) {
+                addTononFinalMethodsCalledOnParam(getClassDescriptorOperand(), getXMethodOperand(), object);
+            }
+        } else if (seen == Const.INVOKESTATIC && getXMethodOperand() != null
+                && "doPrivileged".equals(getXMethodOperand().getName())) {
+            OpcodeStack.Item action = stack.getStackItem(0);
+            CallPair callPair = lookForCalledOutsideAndInside(action);
+            if (callPair != null) {
+                reportBug(callPair);
+            }
+        }
+    }
+
+    private boolean isNestingMethodLocalVariable(OpcodeStack.Item object) {
+        XField field = object.getXField();
+        if (field == null) {
+            return false;
+        }
+
+        return field.getName().matches("val\\$.*");
+    }
+
+    private void addToMethodsCalledInsidePrivilegedAction(XMethod calledMethod, OpcodeStack.Item object) {
+        Set<CallerInfo> objects = methodsCalledInsidePrivilegedAction.get(calledMethod);
+        if (objects == null) {
+            objects = new HashSet<>();
+            methodsCalledInsidePrivilegedAction.put(calledMethod, objects);
+        }
```

Comment:
The [`Map#computeIfAbsent(Function)`](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-) is better replacement to shorten the code:

```suggestion
        Set<CallerInfo> objects = methodsCalledInsidePrivilegedAction.computeIfAbsent(calledMethod, HashSet::new);
```

Reply:
I had to use `k -> new HashSet<>()` because `HashSet::new` did not compile: javac could not deduce CallerInfo as the generic parameter of the HashSet. HashSet<>::new was syntactically invalid and there was also a problem using HashSet<CallerInfo>::new.